src/plan_expand_hypertable.c

```diff
@@ -51,54 +92,67 @@ collect_quals_walker(Node *node, CollectQualCtx *ctx)
 		{
 			Node	   *qual = (Node *) lfirst(lc);
 			RestrictInfo *restrictinfo;
+			Relids		relids;
+			bool		belongs_to_cur_rel = false;
 
-			Relids		relids = pull_varnos(qual);
+			relids = pull_varnos(qual);
 
-			if (bms_num_members(relids) != 1 || !bms_is_member(ctx->rel->relid, relids))
+			belongs_to_cur_rel = bms_num_members(relids) == 1 && bms_is_member(ctx->rel->relid, relids);
+
+			if (!belongs_to_cur_rel)
 				continue;
+
+			if (is_chunk_exclusion_func(qual))
+			{
+				FuncExpr   *func_expr = (FuncExpr *) qual;
+
+				/* validation */
+				if (ctx->chunk_exclusion_func != NULL && belongs_to_cur_rel)
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("only one chunks_in call is allowed per hypertable")));
+
+				Assert(func_expr->args->length == 2);
+				if (!IsA(linitial(func_expr->args), Var))
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("first parameter for chunks_in function needs to be a relation or alias")));
+
+				if (belongs_to_cur_rel)
+					ctx->chunk_exclusion_func = func_expr;
+			}
+
+			if (ctx->chunk_exclusion_func == NULL)
+			{
 #if PG96
-			restrictinfo = make_restrictinfo((Expr *) qual,
-											 true,
-											 false,
-											 false,
-											 relids,
-											 NULL,
-											 NULL);
+				restrictinfo = make_restrictinfo((Expr *) qual,
+												 true,
+												 false,
+												 false,
+												 relids,
+												 NULL,
+												 NULL);
 #else
-			restrictinfo = make_restrictinfo((Expr *) qual,
-											 true,
-											 false,
-											 false,
-											 ctx->root->qual_security_level,
-											 relids,
-											 NULL,
-											 NULL);
+				restrictinfo = make_restrictinfo((Expr *) qual,
+												 true,
+												 false,
+												 false,
+												 ctx->root->qual_security_level,
+												 relids,
+												 NULL,
+												 NULL);
 #endif
-			ctx->result = lappend(ctx->result, restrictinfo);
+				ctx->restrictions = lappend(ctx->restrictions, restrictinfo);
+			}
+		}
+		if (ctx->chunk_exclusion_func != NULL)
+		{
+			f->quals = (Node *) list_delete((List *) f->quals, ctx->chunk_exclusion_func);
```

Comment:
Although I don't think that's strictly necessary in the context of the current PR and existing code, I do agree that it will make things more future proof. I've added utility method for detaching list items without deallocation memory.